libraries/anvill_passes/include/anvill/SliceManager.h

```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#include <llvm/ADT/ArrayRef.h>
+#include <llvm/ADT/SmallVector.h>
+#include <llvm/IR/LLVMContext.h>
+#include <llvm/IR/Module.h>
+#include <llvm/Transforms/Utils/ValueMapper.h>
+
+#include <memory>
+#include <tuple>
+
+#include "SliceInterpreter.h"
+
+namespace anvill {
+
+struct ClonedInstructions {
+  llvm::ValueMap<llvm::Value *, llvm::Value *> mappingOldToNew;
+  llvm::SmallVector<llvm::Instruction *> copied;
+};
+
+class SliceManager;
+class SliceID {
+  friend class SliceManager;
+
+ private:
+  uint64_t ID;
+
+  SliceID() : ID(0) {}
+
+  SliceID operator++(int) {
+    auto temp = *this;
+    this->ID++;
+    return temp;
+  }
+};
+class SliceInterpreter;
+class SliceManager {
+
+ public:
+  class Slice {
+   private:
+    llvm::Function *reprFunction;
+    SliceID ID;
+    // we need origin info for arguments somehow to basically allow analyses to get more context for the slice if they fail.
+   public:
+    Slice(llvm::Function *f, SliceID id) : reprFunction(f), ID(id) {}
+
+    llvm::Function *getRepr() {
+      return this->reprFunction;
+    }
+  };
+
+
+ protected:
+  llvm::LLVMContext context;  // unique context?
+  std::unique_ptr<llvm::Module> mod;
+
+ private:
+  // perhaps at some point we should split modules
+
+  SliceID nextID;
+  std::map<uint64_t, Slice> slices;
+
+  llvm::SmallVector<llvm::Instruction *>
+  createMapperFromSlice(llvm::ArrayRef<llvm::Instruction *> slice,
+                        llvm::ValueToValueMapTy &mapper);
+
+  void insertClonedSliceIntoFunction(llvm::Function *targetFunc,
+                                     llvm::Value *newRet,
+                                     llvm::ArrayRef<llvm::Instruction *> slice);
+
+  std::string getNextFunctionName();
+  llvm::Function *
+  createFunctionForCurrentID(llvm::ArrayRef<llvm::Value *> arguments,
+                             llvm::Value *returnVal);
+
+
+ public:
+  static std::string getFunctionName(SliceID id) {
+    return "sliceFunc." + std::to_string(id.ID);
+  }
+
+  /**
+             * @brief Adds a slice of instructions to the slice manager. Any values not defined in the slice are lifted to arguments.
```

Comment:
Why is this so indented? Perhaps the result of clang-format, or a side-effect my random post-processing script?

Reply:
We generally just end up using `//` comments everywhere.